third_party/nvidia/lib/Dialect/NVWS/Transforms/LowerAref.cpp

```diff
@@ -55,294 +56,587 @@ namespace triton {
-    auto loc = op.getLoc();
-    auto ctx = op.getContext();
-    auto uses = op.getResult().getUses();
-    auto numBatches = op.getResult().getType().getNumBatchAxes();
-
-    int rank = 0;
-    if (numBatches)
-      rank = *numBatches;
-    if (rank > 1)
-      op.emitError("TODO: Implement multi-axis slicing");
-
-    int depth = 1;
-    if (rank == 1) {
-      if (auto mType = dyn_cast<MemDescType>(op.getOperand(0).getType()))
-        depth = mType.getShape()[0];
-      if (auto rType = dyn_cast<RankedTensorType>(op.getOperand(0).getType()))
-        depth = rType.getShape()[0];
-    }
+ArefValue createAndInitMbar(ArefCreateOp op, PatternRewriter &rewriter) {
+  auto [producerArrivalCount, consumerArrivalCount] = getArrivalCount(op);
 
-    rewriter.setInsertionPointAfter(op);
-    Attribute sharedMemorySpace = triton::gpu::SharedMemorySpaceAttr::get(ctx);
-    auto barrierCTALayout = CTALayoutAttr::get(
-        /*context=*/ctx, /*CTAsPerCGA=*/{1},
-        /*CTASplitNum=*/{1}, /*CTAOrder=*/{0});
-    auto barrierEncoding =
-        SwizzledSharedEncodingAttr::get(ctx, 1, 1, 1, {0}, barrierCTALayout);
-    auto barrierType = MemDescType::get(
-        SmallVector<int64_t>{depth}, rewriter.getI64Type(), barrierEncoding,
-        sharedMemorySpace, /*mutableMemory=*/true);
-    // Create two mbarriers
-    auto emptyMbars = rewriter.create<LocalAllocOp>(loc, barrierType, Value());
-    emptyMbars->setAttr("aref_empty_mbarriers", rewriter.getUnitAttr());
-
-    auto fullMbars = rewriter.create<LocalAllocOp>(loc, barrierType, Value());
-    fullMbars->setAttr("aref_full_mbarriers", rewriter.getUnitAttr());
-
-    auto lb = rewriter.create<arith::ConstantIntOp>(loc, 0, 32);
-    auto ub = rewriter.create<arith::ConstantIntOp>(loc, depth, 32);
-    auto step = rewriter.create<arith::ConstantIntOp>(loc, 1, 32);
-    auto dLoop = rewriter.create<scf::ForOp>(loc, lb, ub, step);
-    rewriter.setInsertionPointToStart(dLoop.getBody());
-
-    for (int i = 0; i < 2; ++i) {
-      auto memDesc = getBarrierMemDesc(rewriter, {1});
-      auto singleBarrier = rewriter.create<triton::gpu::MemDescSubviewOp>(
-          loc, memDesc, i == 0 ? emptyMbars.getResult() : fullMbars.getResult(),
-          SmallVector<Value>{dLoop.getInductionVar()});
-      int count = i == 0 ? 0 : 1;
-      rewriter.create<InitBarrierOp>(loc, singleBarrier, count);
-    }
+  MLIRContext *ctx = op.getContext();
+  auto loc = op.getLoc();
+  auto arefTy = op.getType();
+  auto baseType = arefTy.getBaseType();
+  auto arefBufTypes = llvm::to_vector(llvm::map_range(
+      arefTy.getBaseType(), [](Type type) { return cast<MemDescType>(type); }));
+  auto shape = arefBufTypes[0].getShape();
+  auto depth = shape[0];
+
+  MemDescType barrierMemDescType = getBarrierMemDesc(rewriter, {depth});
+  // Create two mbarriers
+  auto emptyMbars =
+      rewriter.create<LocalAllocOp>(loc, barrierMemDescType, Value());
+  emptyMbars->setAttr("aref_empty_mbarriers", rewriter.getUnitAttr());
```

Comment:
why do these ops need to be annotated?

Reply:
same as above https://github.com/triton-lang/triton/pull/7479#discussion_r2205987827, the only annotation that has functional effect is `first_get` attribute on `aref_create` op, it is lit-tested as well. This is because put/get change meaning to "ping/pong" for `aref-tmem-insertion`, and depending on communication pattern, and aref "handshaking", and aref maybe created with `get` being first, rather than `put`, and we need to have it associated mbarrier be in released state, rather than that of `put`, so we `xor` phase of `get` rather than `put`.

Reply:
Randomly adding attributes to ops is not standard MLIR practice. If you want to add these attributes solely for the purpose of writing lit tests, the addition of the attributes should be guarded behind a pass option that is only set by the lit test.

Reply:
That being said, I am not convinced you really need attributes to effectively lit test the IR in this case.

Reply:
> the only annotation that has functional effect is first_get attribute on aref_create op

I don't fully understand the functional effects of `first_get`. Can you give a few examples?